src/RMPI/mpidata.jl

```diff
@@ -217,13 +217,57 @@ end
 function LinearAlgebra.dot(x, md::MPIData)
     return MPI.Allreduce(localpart(x)⋅localpart(md), +, md.comm)
 end
-function LinearAlgebra.dot(x, lop, md::MPIData)
-    return MPI.Allreduce(dot(x, lop, localpart(md)), +, md.comm)
-end
```

Comment:
Don't we need this method? Say we want to compute the projected energy for a Hamiltonian for which the adjoint is not known. Then `x` is a short `DVec` that is copied to all ranks and the overlaps can be computed locally and then reduced. Sending spawns around via MPI is overkill and not required. 

Maybe we restrict the type of `x` to be not `MPIData` (Is that possible?) in order to make this clear for readers of the code and in order to avoid wrong results if the other method does not dispatch as it should.

Reply:
Good catch. I don't think we can restrict the type like that, but having the method definitions close to each other will make it clear enough.

Reply:
As we discussed, a comment explaining that `x` is expected to be local and not MPI distributed would be helpful.
What about something like
```
    @assert x ≡ localpart(x) "Local method for `dot` called with non-local argument"
``` 
?
Also, we discussed adding a method for `dot(md::MPIData, lop, x)`

Reply:
I added the comments and a few new methods. Now any combination of `MPIData` and regular data should work, assuming `MPIData`s are distributed correctly and local vectors are truly local.